### PR TITLE
Space - Stephanie

### PR DESCRIPTION
**Pull request by stephaniejmars**
# Assignment Submission: Hotel

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | My design definitely evolved. I was excited to do my own design, but whenever I had to ask for help with something people I asked  would question my design, and sometimes it was helpful, but sometimes it was a waste of time, and I wasn't getting the help I needed on simple syntax type issues I was having. 
What was a design decision you made that changed over time over the project? | I wanted Reservation to build reservations, but in the end I decided it would be a reservation and not have a reservation. Then I decided it would go in Manager, and eventually making a reservation ended up in Room.
What was a concept you gained clarity on, or a learning that you'd like to share? | I learned a little more about Has vs Is. I still need clarity on attr_reader, Let, testing pieces within methods instead of just the outcome. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | I tested if a room is available on a date that was months away from any of its bookings.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | I tested if a room is booked 4/1-4/2 and 4/3-4/7, it should be available to book 4/2-4/3 (and it wasn't, so I fixed it).
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | It is hard at first when you don't funny understand the syntax of the test, but it was helpful to do just small pieces of code and test at a time. Make sure its passing before I moved on.

### CODE REVIEW

File: lib/manager.rb [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+require 'date'
+require_relative 'room'
+require_relative 'reservation'
+
+
+module Hotel
+  class Manager
+    attr_reader :num, :all_rooms, :rm_reservations, :recloc
+    attr_writer
+    attr_accessor
```
Comment (jmaddox19) [inline lib/manager.rb:11]: Just so you know, these lines aren't needed when you don't have any attributes you want to list for them.

File: lib/manager.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+require_relative 'reservation'
+
+
+module Hotel
+  class Manager
+    attr_reader :num, :all_rooms, :rm_reservations, :recloc
+    attr_writer
+    attr_accessor    
+
+    def initialize(num: 20)
```
Comment (jmaddox19) [inline lib/manager.rb:13]: This is a great way to set this up! A great example of where input params to a constructor can be totally different than the instance variable(s)!

File: lib/manager.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      if num == nil || num == 20 
+        @all_rooms = Manager.create_rooms
+      else
+        @all_rooms = Manager.create_rooms(num: num)
+      end
+    end
+
+    def self.create_rooms(num: 20)
```
Comment (jmaddox19) [inline lib/manager.rb:21]: Cool that you made this its own method!

File: lib/manager.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        if room.is_available_range(start_date, end_date) == true 
+          return room.book_room(start_date, end_date)
+        end 
+      end
+      raise ArgumentError.new("No rooms available")
+    end
+
+    def res_by_room(rm_num, start_date, end_date)
+      res_array = []
+      @found_room = @all_rooms.find {|room| room.rm_num}
```
Comment (jmaddox19) [inline lib/manager.rb:50]: This could/should be a local variable rather than an instance variable since it is only used within this one method.
That would be changed by just removing the `@`
```suggestion
      found_room = @all_rooms.find {|room| room.rm_num}
```

File: lib/manager.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
+require 'time'
+require 'date'
+require_relative 'room'
+require_relative 'reservation'
+
+
+module Hotel
+  class Manager
+    attr_reader :num, :all_rooms, :rm_reservations, :recloc
```
Comment (jmaddox19) [inline lib/manager.rb:9]: This should only define readers for instance variable of the `Hotel` class. The only of these 4 that is an instance variable of `Hotel` is `all_rooms`. 
```suggestion
    attr_reader :all_rooms
```

File: lib/reservation.rb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+module Hotel
+  class Reservation
+    attr_reader :start_date, :end_date, :rm_num, :cost_per_day, :total_cost, :recloc
+
+    def initialize(start_date:,
+      end_date:, 
+      rm_num: nil, 
+      cost_per_day: 200, 
+      total_cost: nil, 
+      recloc: nil)
```
Comment (jmaddox19) [inline lib/reservation.rb:16]: I'm having trouble understanding what this variable represents. It's best to name variables in such a way that their purpose is clear to anyone who might read the code.

File: lib/reservation.rb [hunk 7]
```diff
@@ -0,0 +1,32 @@
+require 'time'
+require 'date'
+require 'securerandom'
+require_relative 'manager'
+require_relative 'room'
+
+module Hotel
+  class Reservation
+    attr_reader :start_date, :end_date, :rm_num, :cost_per_day, :total_cost, :recloc
```
Comment (jmaddox19) [inline lib/reservation.rb:9]: These all correspond to instance variables. Great!

File: lib/manager.rb (hunk 5 above)
Comment (jmaddox19) [reply lib/manager.rb:9]: I imagine you might understand that and just forgot to update this line.

File: lib/manager.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+          room_array << room
+        end 
+      end
+      return room_array 
+    end
+
+    def find_total_cost(rec_loc)
+      res = ""
+      @all_rooms.each do |room|
+         if room.rm_reservations.find{|res| res.recloc[:recloc] == rec_loc}.class == Hotel::Reservation
```
Comment (jmaddox19) [inline lib/manager.rb:82]: This is just slightly too indented. 
I use this handy VSCode plugin called "indent-rainbow" to make that really easy to notice.

**Comment by jmaddox19** [comment]
# Hotel






## Section 1: Major Learning Goals



| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
Practices SRP by having at least two separate classes with distinct responsibilities, and test files for these two classes | ✔️
Overall, demonstrates understanding instance variables vs. local variables. (There aren't unnecessarily too many instance variables, when it should be a local variable) | There seem to be some possible gaps in understanding here. See my inline comments for more details.
For each test file, tests demonstrate an understanding of instantiating objects properly, and using Arrange-Act-Assert | ✔️
Practices pseudocode and TDD, and reflected on it by filling out the reflection questions | ✔️
Practices git with at least 15 small commits and meaningful commit messages | ✔️

## Section 2: Code Review and Testing Requirements

| Criteria | yes/no, and optionally any details/lines of code to reference |
| --- | --- |
There is a class that represents a reservation, and a second class that holds/manages a collection of reservations through composition (instance variable) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date is complex logic that is separated into method(s) (and potentially class(es)) | ✔️
The logic for checking if a reservation's date overlaps with another reservation's date has unit tests | ✔️
All of the given tests run and pass | ✔️
A test coverage tool is installed and used, and shows 95% test coverage | ✔️ 99%!!

## Section 3: Feature Requirements

| Feature Requirement: There is a method that... | yes/no |
| --- | --- |
gives back a list of rooms, and it's tested | ✔️
creates a specific reservation for a room for a given date range, and it has nominal test cases | ✔️
creates a specific reservation for a room for a given date range, and it tests an edge case, such as no available room, or invalid date range | ✔️
gives back a list of reservations on a given date. Its tests include a test that makes several reservations for a given date | ✔️
calculates the total price for a reservation | ✔️
gives back a list of available rooms for a given date range, and it has nominal test cases | ✔️
gives back a list of available rooms for a given date range, and it has edge test cases, such as no available room, or invalid date range | 
creates a block of rooms | 
reserves a room from a block | 

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ total in all sections | ✔️
| Yellow (Approaches Standards) | 9-13 total in all sections |
| Red (Not at Standard) | 0-8 total in all sections, or assignment is breaking/doesn’t run with less than 5 minutes of debugging |

### Additional Feedback



Great work overall! You've built your first project with minimal starting code. This represents an incredible milestone in your journey, and you should be proud of yourself!

I am particularly impressed by the way that you setup readable and logical relationships between classes!

I do see some room for improvement around variable naming, instance variables, and readers/writers. Please see the inline comments for more details.


Please reach out to me if any of my comments don't give enough detail or if you could use more explanation!

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | 
| Elegant/Clever | 
| Descriptive/Readable | ✅
| Concise | 
| Logical/Organized | ✅
